File: 백준/Platinum/4384. 공평하게 팀 나누기/공평하게 팀 나누기.java


import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);



	
	

			int n = scan.nextInt();
			int[] w = new int[n+1];
			int sum = 0;

			for(int i = 1; i <= n; i++) {
				w[i] = scan.nextInt();
				sum += w[i];
			}

			int[][] dp = new int[n+1][sum+1];

			for(int i = 1; i <= n; i++) {
				dp[i][w[i]] = 1;
				for(int j = 1; j <= sum; j++) {
					if(j > w[i]) {
						for(int index=1;index<i;index++)
			                   if(dp[index][j-w[i]]>0)
			                   {
			                      dp[i][j] = dp[index][j-w[i]]+1;
			                   }
					}
				}
			}

			int min = sum;
			int ans = sum;
			for(int i=1;i<=n;i++)
			{
				for(int j = 1; j <= sum; j++) 
				{
					if(dp[i][j]==n/2) 
					{
						if( Math.abs(2*j - sum) < min)
						{
							min = Math.abs(sum-2*j);
							ans = j;
						}
					}
					else if(n%2==1 && (dp[i][j]==n/2||dp[i][j]==n/2+1)) 
					{
						if( Math.abs(sum-2*j) < min)
						{
							min = Math.abs(sum-2*j);
							ans = j;
						}
					}
				}
			}
			System.out.println(Math.min(ans,sum-ans)+" "+Math.max(ans, sum-ans));



	}
}

File: 백준/Platinum/4384. 공평하게 팀 나누기/공평하게 팀 나누기.py
N = int(input())

W = [0]

for _ in range(N):
    W.append(int(input()))

sum = 0

for i in W:
    sum += i

dp = [[0] * (sum+1) for _ in range(N+1)]

for i in range(1, len(W)):
    dp[i][W[i]] = 1
    for j in range(1, sum+1):
        if j > W[i]:
            for idx in range(1, i):
                if dp[idx][j-W[i]] > 0:
                    dp[i][j] = dp[idx][j-W[i]] + 1

minW = sum
result = sum

for i in range(1, N+1):
    for j in range(1, sum+1):
        if dp[i][j] == N // 2:
            if abs(2*j - sum) < minW:
                minW = abs(2 * j - sum)
                result = j

print(min(result, abs(result-sum)))
print(max(result, abs(result-sum)))

